### ISSUE TITLE
nccl-test Indicates a performance problem

### ISSUE THREAD

**Issue by yalbaba** [NVIDIA/nccl#1511]
Scheduling two containers on the same node results in significantly lower performance for nccl-test than scheduling two containers on different nodes

An experiment to schedule two containers to different nodes：
![Image](https://github.com/user-attachments/assets/d706e47c-dba4-4525-9770-4b0217da7d85)

An experiment to schedule two containers to the same node：
![Image](https://github.com/user-attachments/assets/70065b94-9fd8-4613-8b3d-6c5173947e30)

Under normal circumstances, the gpu communication on the same node is through the nvlink channel, but why the bus bandwidth between the GPU of the container on the same node is low？

**Comment by kiskra-nvidia** [comment]
Hard to say what might be going on from just these two screenshots. I recommend running both configs with `NCCL_DEBUG=INFO NCCL_DEBUG_SUBSYS=INIT,ENV,GRAPH` and posting the output files here -- that should give us a lot more details about what connectivity NCCL sees between various devices.

**Comment by yalbaba** [comment]
This is my output file:
[outeput.txt](https://github.com/user-attachments/files/17694396/outeput.txt)

**Comment by yalbaba** [comment]
This is my output file：

**Comment by yalbaba** [comment]
This is my output file：
[outeput.txt](https://github.com/user-attachments/files/17694405/outeput.txt)

The experimental background is: a single node starts two containers to test nccl-test

**Comment by yalbaba** [comment]
> Hard to say what might be going on from just these two screenshots. I recommend running both configs with `NCCL_DEBUG=INFO NCCL_DEBUG_SUBSYS=INIT,ENV,GRAPH` and posting the output files here -- that should give us a lot more details about what connectivity NCCL sees between various devices.

This is my output file：
[outeput.txt](https://github.com/user-attachments/files/17694412/outeput.txt)


The experimental background is: a single node starts two containers to test nccl-test

**Comment by kiskra-nvidia** [comment]
OK, so the above is the output when running the two containers on the same node? Can you post the same output when running each container on a different node?

What is your end goal here? When splitting GPUs between two containers, NCCL naturally thinks that it's running on two different nodes so it uses the NICs rather than the NVlinks to communicate between the GPUs that are not in the same container. I'm assuming that `all_reduce_perf` performs considerably faster on your system when running bare-metal (or, at least, when all 4 GPUs are within the same container)? Have you tried that? Still, I agree that it's still an interesting question why things run considerably faster when the two containers are on two different physical nodes...

**Comment by yalbaba** [comment]
> OK, so the above is the output when running the two containers on the same node? Can you post the same output when running each container on a different node?
> 
> What is your end goal here? When splitting GPUs between two containers, NCCL naturally thinks that it's running on two different nodes so it uses the NICs rather than the NVlinks to communicate between the GPUs that are not in the same container. I'm assuming that `all_reduce_perf` performs considerably faster on your system when running bare-metal (or, at least, when all 4 GPUs are within the same container)? Have you tried that? Still, I agree that it's still an interesting question why things run considerably faster when the two containers are on two different physical nodes...

My goal is to ensure that the communication speed between Gpus starting two containers on the same node is the same as the communication speed between Gpus starting two containers on different nodes, or even faster. I believe that since nccl has the ability to automatically sense the hardware topology, Gpus on the same physical machine should use nvlink to communicate, at least not more slowly than across physical machines.

**Comment by yalbaba** [comment]
> OK, so the above is the output when running the two containers on the same node? Can you post the same output when running each container on a different node?
> 
> What is your end goal here? When splitting GPUs between two containers, NCCL naturally thinks that it's running on two different nodes so it uses the NICs rather than the NVlinks to communicate between the GPUs that are not in the same container. I'm assuming that `all_reduce_perf` performs considerably faster on your system when running bare-metal (or, at least, when all 4 GPUs are within the same container)? Have you tried that? Still, I agree that it's still an interesting question why things run considerably faster when the two containers are on two different physical nodes...

This is the topology of my two containers：
![Image](https://github.com/user-attachments/assets/d38c74e1-6806-4b3a-ad30-b45564238380)

![Image](https://github.com/user-attachments/assets/57a4df2c-237a-4b4b-bfaf-0892e0615518)

But when I use nccl, I can't use the nvlink channel:
![Image](https://github.com/user-attachments/assets/47e12e30-e878-439d-8ffb-c654194f7b9f)

**Comment by kiskra-nvidia** [comment]
> But when I use nccl, I can't use the nvlink channel: ![Image](https://github.com/user-attachments/assets/47e12e30-e878-439d-8ffb-c654194f7b9f)

Are you referring to the `0 nvls channels` part? That doesn't mean that NVLink is not being used. NVLS (NVLink SHARP) is a feature of the fourth-generation NVLink (starting with Hopper; you have Ampere).

**Comment by sjeaugey** [comment]
> Can you go into more detail about this, are you saying that Ampere's GPU can't use nvlink?

Ampere can use NVLink. Just not NVLink SHARP.

**Comment by kiskra-nvidia** [comment]
Did you convert the topology file from XML to JSON?

If I understand it correctly, your latest `mpirun` command line runs `all_reduce_perf` on 4 GPUs total (2 per container), correct? If this corresponds to the NCCL debug output from https://github.com/user-attachments/files/17694412/outeput.txt, then I'm guessing that NVLink _is_ being used for communication between the two GPUs within each container; what makes you think that it's not?

**Comment by yalbaba** [comment]
> Did you convert the topology file from XML to JSON?
> 
> If I understand it correctly, your latest `mpirun` command line runs `all_reduce_perf` on 4 GPUs total (2 per container), correct? If this corresponds to the NCCL debug output from https://github.com/user-attachments/files/17694412/outeput.txt, then I'm guessing that NVLink _is_ being used for communication between the two GPUs within each container; what makes you think that it's not?

Yes, I converted the topology into json format. I have two questions. First, why does the topology show that the two GPUs are under the same Numa node, and I set NCCL_P2P_LEVEL=PHB, but NVLINK is not used? (I think the reason why NVLINK is not used is that the bandwidth shown in the test results is too small, far from the speed of NVLINK). Second, why I tested NCCL in two containers, and each container used 2 GPUs, but only 2 GPU nodes were shown in the topology?

**Comment by kiskra-nvidia** [comment]
The NUMA topology is important when GPUs communicate over PCIe, but when they are connected using NVLink (as in your case) then the topology shouldn't matter. Looking back at your `outeput.txt` file from https://github.com/NVIDIA/nccl/issues/1511#issuecomment-2467110220, the following lines indicate that P2P _is_ being used, presumably using NVLink not PCIe:

```
nccl-test-545c9b9fd4-dvp4k:188:196 [1] NCCL INFO GPU/4B000 :GPU/4B000 (0/5000.000000/LOC) GPU/51000 (2/160.000000/NVL) NVS/0 (1/160.000000/NVL) CPU/0 (3/24.000000/PHB) CPU/1 (4/10.000000/SYS) NET/0 (4/12.500000/PXB) NET/1 (4/12.500000/PXB) NET/2 (6/12.500000/PHB) NET/3 (6/12.500000/PHB) NET/4 (7/10.000000/SYS) NET/5 (7/10.000000/SYS) NET/6 (7/10.000000/SYS) NET/7 (7/10.000000/SYS) 
nccl-test-545c9b9fd4-dvp4k:188:196 [1] NCCL INFO GPU/51000 :GPU/4B000 (2/160.000000/NVL) GPU/51000 (0/5000.000000/LOC) NVS/0 (1/160.000000/NVL) CPU/0 (3/24.000000/PHB) CPU/1 (4/10.000000/SYS) NET/0 (4/12.500000/PXB) NET/1 (4/12.500000/PXB) NET/2 (6/12.500000/PHB) NET/3 (6/12.500000/PHB) NET/4 (7/10.000000/SYS) NET/5 (7/10.000000/SYS) NET/6 (7/10.000000/SYS) NET/7 (7/10.000000/SYS) 
[...]
nccl-test-545c9b9fd4-fhnm4:75:81 [0] NCCL INFO Channel 00/0 : 2[0] -> 3[1] via P2P/direct pointer/read
[...]
nccl-test-545c9b9fd4-dvp4k:188:195 [0] NCCL INFO Channel 00/0 : 0[0] -> 1[1] via P2P/direct pointer/read
```

If you want to verify it, then the experiment to run would be to use just _one_ container (with two GPUs) and then compare the performance you get with `NCCL_P2P_LEVEL=LOC` (P2P disabled) to the one without.

> Second, why I tested NCCL in two containers, and each container used 2 GPUs, but only 2 GPU nodes were shown in the topology?

The topology file shows the topology from the point of view of a single NCCL rank, normally `0`, and it includes local resources only. You can change which rank generates the topology file via `NCCL_TOPO_DUMP_FILE_RANK=<rank>`, but I don't expect the output to be significantly different...
